docker/docker_image_src.go

```diff
@@ -30,15 +32,68 @@ type dockerImageSource struct {
 
 // newImageSource creates a new ImageSource for the specified image reference.
 // The caller must call .Close() on the returned ImageSource.
-func newImageSource(sys *types.SystemContext, ref dockerReference) (*dockerImageSource, error) {
-	c, err := newDockerClientFromRef(sys, ref, false, "pull")
+func newImageSource(ctx context.Context, sys *types.SystemContext, ref dockerReference) (*dockerImageSource, error) {
+	registry, err := sysregistriesv2.FindRegistry(sys, ref.ref.Name())
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "error loading registries")
+	}
+
+	if registry == nil {
+		// No registry was found be the provided reference, so we create a
+		// fallback registry by hand to make the client creation below work as
+		// intended.
+		registry = &sysregistriesv2.Registry{
+			Endpoint: sysregistriesv2.Endpoint{
+				URL: ref.ref.String(),
+			},
+		}
+	}
+
+	// Found the registry within the sysregistriesv2 configuration. Now we test
+	// all endpoints for the manifest availability. If an working image source
+	// was found, it will be used for all future pull actions.
+	var (
+		imageSource     *dockerImageSource
+		manifestLoadErr error
+	)
+	for _, endpoint := range append(registry.Mirrors, registry.Endpoint) {
+		newRef, err := rewriteReference(ref, registry.Prefix, endpoint.URL)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error rewriting reference")
+		}
+
+		client, err := newDockerClientFromRef(sys, newRef, false, "pull")
+		client.endpointSkipVerify = endpoint.Insecure
+		if err != nil {
+			return nil, err
+		}
+
+		testImageSource := &dockerImageSource{
+			ref: newRef,
+			c:   client,
+		}
+
+		manifestLoadErr = testImageSource.ensureManifestIsLoaded(ctx)
+		if manifestLoadErr == nil {
+			imageSource = testImageSource
+			break
+		}
 	}
-	return &dockerImageSource{
-		ref: ref,
-		c:   c,
-	}, nil
+
+	return imageSource, manifestLoadErr
+}
+
+// rewriteReference will substitute the provided reference prefix `from` to the
+// given `to` from the `ref` and creates a new reference from it. The function
+// errors if the newly created reference is not parsable.
+func rewriteReference(ref dockerReference, from, to string) (dockerReference, error) {
+	stringRef := ref.ref.String()
+	if from != "" && strings.HasPrefix(stringRef, from) {
```

Comment:
We need a debug log here. I am testing the Skopeo PR and desperately want to know what's going under the hood :^)

Reply:
Adapted as suggested by both of you.

Reply:
> he fallback value instead

The major issue is that it is not that trivial to parse the prefix from the reference and I would keep that functionality within the sysregistries implementation. I prefer to stick to the current behavior: If no prefix is provided, simply return the reference. I added this as an early return path to make it more clear. Alternatively It would be possible to add something like "fromRef(ref reference.Named) Registry` to sysregistries but this seems out of scope to me.

Reply:
> The major issue is that it is not that trivial to parse the prefix from the reference and I would keep that functionality within the sysregistries implementation.

```go
Endpoint: sysregistriesv2.Endpoint{
	URL: ref.ref.String(),
	prefix: ref.ref.String(),
},
```
should, AFAICS, work just fine.

Reply:
Fix proposed in https://github.com/containers/image/pull/617 .